adc/read_vsys/power_status.c

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright (c) 2023 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "stdbool.h"
+#include "hardware/adc.h"
+#include "power_status.h"
+
+#if CYW43_USES_VSYS_PIN
+#include "pico/cyw43_arch.h"
+#endif
+
+#ifndef PICO_POWER_SAMPLE_COUNT
+#define PICO_POWER_SAMPLE_COUNT 3
+#endif
+
+int power_battery(bool *battery_powered) {
+#if defined CYW43_WL_GPIO_VBUS_PIN
+    *battery_powered = !cyw43_arch_gpio_get(CYW43_WL_GPIO_VBUS_PIN);
+    return PICO_OK;
+#elif defined PICO_VBUS_PIN
+    gpio_set_function(PICO_VBUS_GPIO_PIN, GPIO_FUNC_SIO);
+    *battery_powered = !gpio_get(PICO_VBUS_GPIO_PIN);
+    return PICO_OK;
+#else
+    return PICO_ERROR_NO_DATA;
+#endif
+}
+
+int power_voltage(float *voltage_result) {
+#ifndef PICO_VSYS_PIN
+    return PICO_ERROR_NO_DATA;
+#endif
+#ifndef PICO_VSYS_ADC_INPUT
+    #error If PICO_VSYS_PIN is defined PICO_VSYS_ADC_INPUT must also be defined
```

Comment:
This error message _seems_ to imply that this second `#ifndef` ought to be nested inside the previous `#ifndef` ?